### PR TITLE
Feature/caching with api connection info

### PR DESCRIPTION
**Pull request by rma-rripken**
Based on Bryson's PR #117 

I merged main and then added cache to ApiConnectionInfo.

### CODE REVIEW

File: cwms-http-client/src/main/java/mil/army/usace/hec/cwms/http/client/CwmsHttpCache.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+            this.cacheSize = cacheSize;
+            return this;
+        }
+
+        public Cache build(){
+            try {
+                Files.createDirectories(cacheDirectory);
+                return new Cache(cacheDirectory.toFile(), cacheSize * 1024 * 1024);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
```
Comment (adamkorynta) [inline cwms-http-client/src/main/java/mil/army/usace/hec/cwms/http/client/CwmsHttpCache.java:49]: I think `IllegalStateException` is more appropriate and should have a message context with which directory was throwing the error

File: cwms-http-client/src/main/java/mil/army/usace/hec/cwms/http/client/CwmsHttpCache.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+
+    static Cache getInstance() {
+        return INSTANCE;
+    }
+
+    public static class Builder{
+        private Path cacheDirectory = Paths.get(System.getProperty(CACHE_DIRECTORY_PROPERTY_KEY, CACHE_DEFAULT_DIRECTORY.toString()));
+        private long cacheSize = Long.parseLong(System.getProperty(CACHE_SIZE_PROPERTY_KEY, CACHE_DEFAULT_SIZE_MB + ""));
```
Comment (adamkorynta) [inline cwms-http-client/src/main/java/mil/army/usace/hec/cwms/http/client/CwmsHttpCache.java:29]: track the units in the variable name `cacheSizeMb`

File: cwms-http-client/src/main/java/mil/army/usace/hec/cwms/http/client/CwmsHttpCache.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    private CwmsHttpCache() {
+        throw new AssertionError("Singleton utility class, do not instantiate");
+    }
+
+
+    static Cache getInstance() {
+        return INSTANCE;
+    }
+
+    public static class Builder{
```
Comment (adamkorynta) [inline cwms-http-client/src/main/java/mil/army/usace/hec/cwms/http/client/CwmsHttpCache.java:27]: inner class should be packaged scope to match outer class

File: cwms-http-client/src/main/java/mil/army/usace/hec/cwms/http/client/ApiConnectionInfoBuilder.java [hunk 4]
```diff
@@ -79,6 +81,11 @@ public ApiConnectionInfoBuilder withHostnameVerifier(HostnameVerifier hostnameVe
         return this;
     }
 
+    public ApiConnectionInfoBuilder withCache(Cache cache) {
```
Comment (adamkorynta) [inline cwms-http-client/src/main/java/mil/army/usace/hec/cwms/http/client/ApiConnectionInfoBuilder.java:84]: This is exporting the OkHttp API

Comment (rma-rripken) [reply cwms-http-client/src/main/java/mil/army/usace/hec/cwms/http/client/ApiConnectionInfoBuilder.java:84]: What do you suggest?  A supplier class sort of like the CookieJar uses?  Doesn't that still tie onto the OkHttp classes?

**Review by MikeNeilson — COMMENTED** [review]
Nothing to add beyond what Adam reported at this time.

File: cwms-http-client/src/main/java/mil/army/usace/hec/cwms/http/client/CwmsHttpCache.java (hunk 3 above)
Comment (rma-rripken) [reply cwms-http-client/src/main/java/mil/army/usace/hec/cwms/http/client/CwmsHttpCache.java]: done

File: cwms-http-client/src/main/java/mil/army/usace/hec/cwms/http/client/CwmsHttpCache.java (hunk 1 above)
Comment (rma-rripken) [reply cwms-http-client/src/main/java/mil/army/usace/hec/cwms/http/client/CwmsHttpCache.java]: done

**Review by adamkorynta — APPROVED** [review]

**Review by adamkorynta — APPROVED** [review]

**Pull request merged**